Promote SchedulerPreemption e2e tests to Conformance

spiffxp opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Promotes the following tests to Conformance:
- `[sig-scheduling] SchedulerPreemption [Serial] validates basic preemption works`
- `[sig-scheduling] SchedulerPreemption [Serial] validates lower priority pod preemption by critical pod`

**Special notes for your reviewer**:

Supercedes https://github.com/kubernetes/kubernetes/pull/78364 which can't merge due to CLA issues

It looked good to myself and other reviewers, I was going to seek `/approve` for this once a rebase had been taken care of (which I'm doing here)

```release-note
NONE
```

/cc @johnbelamaric @bsalamat

spiffxp commented:
/priority important-soon
/milestone v1.16

spiffxp commented:
/assign @bgrant0607 @smarterclayton 
for the /approve

/retest

spiffxp commented:
/override pull-kubernetes-conformance-image-test
failure unrelated https://github.com/kubernetes/kubernetes/issues/80589

BenTheElder commented:
/test pull-kubernetes-e2e-kind

smarterclayton commented:
I was seeing these fail on more resource constrained clusters.  Ravi was looking at it, but I want to

/hold

until we know whether the test is flawed or not.

spiffxp commented:
What is the next actionable step here? Is there an issue we should be watching, job results that can show flakiness? etc.

spiffxp commented:
rebased, @smarterclayton any updates on your end?

spiffxp commented:
anti-affinity test was removed by https://github.com/kubernetes/kubernetes/pull/80821 (and I am fine with complicated scheduling logic tests moving away from e2e)

smarterclayton commented:
@ravisantoshgudimetla can you comment about the gaps in this test that need to be closed for it not to be flaky on resource constrained clusters (the issues we have been seeing with this test running it to test Openshift conformance?)

ravisantoshgudimetla commented:
One issue we noticed with preemption test in OpenShift origin CI is 

- We've hardcoding for CPU(https://github.com/kubernetes/kubernetes/pull/80545/files#diff-e35316a06970f9000b7b154addabda8cL109) and memory (https://github.com/kubernetes/kubernetes/pull/80545/files#diff-e35316a06970f9000b7b154addabda8cL112), this may not hold good in the heavily utilized cluster, either we need to skip the tests when the nodes are that heavily loaded or scale down the values so that they can with minimal resources available on the node.

@damemi and I are actively working on restructuring tests to remove those hardcodings.

/cc @bsalamat

spiffxp commented:
@ravisantoshgudimetla Is there a way we can get the results of heavily utilized clusters to show up on testgrid.k8s.io so we can evaluate the behavior of this and other tests the same way that we do for general flakiness etc?

xmudrii commented:
@spiffxp @smarterclayton Hello! I'm the bug triage lead for the 1.16 release cycle and considering this PR is tagged for 1.16, I'd like to check its status. The code freeze is starting on August 29th EOD PST (about 5 days from now) and once the code freeze starts, only release-blocking PRs will be considered to be merged. Is this PR still planned for the 1.16 release?

smarterclayton commented:
/hold

While we discuss.

On Sun, Aug 25, 2019 at 7:55 AM Marko Mudrinić <notifications@github.com>
wrote:

> @spiffxp <https://github.com/spiffxp> @smarterclayton
> <https://github.com/smarterclayton> Hello! I'm the bug triage lead for
> the 1.16 release cycle and considering this PR is tagged for 1.16, I'd like
> to check its status. The code freeze is starting on August 29th EOD PST
> (about 5 days from now) and once the code freeze starts, only
> release-blocking PRs will be considered to be merged. Is this PR still
> planned for the 1.16 release?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/80545?email_source=notifications&email_token=AAI37J276LGUNJGMZC4HU4TQGJXMDA5CNFSM4IGVVE72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5CSB3I#issuecomment-524624109>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI37JZA22FC4HJM2OFFF73QGJXMDANCNFSM4IGVVE7Q>
> .
>

xmudrii commented:
As the code freeze started, I'm moving it to the 1.17 milestone.

/milestone v1.17

johnbelamaric commented:
@ravisantoshgudimetla @damemi any update here?

smourapina commented:
Friendly reminder that code freeze is rapidly approaching and will take place in November 14th. Is this issue still on track for 1.17? Otherwise we will need to move it to 1.18 milestone.

josiahbjorgaard commented:
/milestone clear

(No activity during 1.17 cycle)
If any concerns please ping me or sig-release on slack.

spiffxp commented:
@smarterclayton @ravisantoshgudimetla @damemi it's been nearly three months, have you addressed your concerns? if not, can we get a checklist of what needs to be changed in case others want to help out here?

damemi commented:
I have been actively working on trying to debug why these tests only fail for openshift in the following PR: https://github.com/openshift/origin/pull/23758

I have made some progress by noticing that for one, the test does not guarantee that the "preemptor pod" will land on the same node as the "victim" pod (which I address [in this commit of that same pr](https://github.com/openshift/origin/pull/23758/commits/11623bc072620238d397e7754f9339f477451555)). I believe this may actually be an error we should fix in k/k as well.

However I cannot figure out an issue this now causes: the preemptor pod lands on the correct node, but fails to run due to low CPU available on that node (and fails to evict the target pod). Is this not the exact case in which the target pod should be evicted, to create more available resources on the node? I am trying to tweak the request calculations in the test but it is baffling that this is not an issue we ever see here, in upstream.

Obviously, this may just be an OpenShift problem. I am simply updating with my findings from the ongoing work to attempt to fix these flakes in high-usage clusters

smarterclayton commented on test/e2e/scheduling/preemption.go:
```diff
@@ -95,10 +95,13 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
+		Description: When a higher priority pod is created and no node with enough resources is found,
+		scheduler MUST preempt a lower priority pod and schedule the high priority pod.
+	*/
+	framework.ConformanceIt("validates basic preemption works", func() {
```
Looking at the code for this test, the calculation to request 40% of allocatable memory is not really a safe check.  It should be subtracting the currently scheduled CPU requests on that node from allocatable, then taking 40%.

damemi replied:
> It should be subtracting the currently scheduled CPU requests on that node from allocatable, then taking 40%.

@smarterclayton in origin, we [do currently do that](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/test/e2e/scheduling/preemption.go#L111-L117), but it still flakes (if you look at my [origin PR above](https://github.com/openshift/origin/pull/23758), I temporarily attempt to revert those changes). I also have had a PR open for a while now in kubernetes to update this as well: https://github.com/kubernetes/kubernetes/pull/82350

smarterclayton commented:
The current logic of this test makes no sense on medium sized nodes which run system critical pods on each node.  The CPU calculation needs to more effectively take into account existing workloads scheduled on that node.  Since increasingly clusters have high priority infrastructure pods running on nodes, the somewhat naive assumptions here aren't appropriate for a conformance test until fixed.

It's unfortunate there is no easy to use API or summary one can get from a test to estimate the available capacity.  I suppose the test will need to estimate it as best it can.

spiffxp commented:
https://github.com/kubernetes/kubernetes/pull/82350 is out to try addressing the concerns mentioned here

hh commented:
@spiffxp needs rebase

spiffxp commented:
I believe https://github.com/kubernetes/kubernetes/pull/82350 addresses @smarterclayton's concerns

Rebased. Tests were refactored and added to, but to avoid blowing scope of this PR I've stuck with the two tests listed in the PR description

spiffxp commented:
/hold cancel
/assign @smarterclayton @johnbelamaric 
for the /approve (and /lgtm)

spiffxp commented:
argh, needs rebase again

spiffxp commented:
two weeks of no flakes: https://testgrid.k8s.io/sig-release-master-informing#gce-cos-master-serial&include-filter-by-regex=validates.*preemption&width=5

spiffxp commented:
/cc @ahg-g @Huang-Wei 
I need someone from sig-scheduling to /lgtm this, emeritus chair bsalamat did it for this PR's earlier form in https://github.com/kubernetes/kubernetes/pull/78364

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -98,10 +98,13 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
-	// enough resources is found, scheduler preempts a lower priority pod to schedule
-	// the high priority pod.
-	ginkgo.It("validates basic preemption works", func() {
+	/*
```
nit: Would you mind correcting the typo in L125: `let not's` -> `let's not`.

spiffxp replied:
done

Huang-Wei commented:
@spiffxp LGTM. Just a pre-existing typo would be good to correct.

Huang-Wei commented:
/lgtm
/approve

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

hh commented:
Ready for review by behaviour approval.

spiffxp commented:
/assign @johnbelamaric @smarterclayton 
Looking for an /approve from conformance approvers

hh commented:
@spiffxp can you rebase?

Riaankl commented:
@spiffxp: PR needs rebase. Can you please support.

spiffxp commented:
It looks like these tests were redone to use an extended resource instead of CPU or memory.  Should be way less flaky now.  Rebasing...

spiffxp committed:
Promote preemption e2e tests to Conformance

spiffxp commented:
/retest

spiffxp commented:
@smarterclayton @johnbelamaric 
looking for /lgtm and /approve here

Huang-Wei commented:
LGTM from the scheduling side. The test is refactored and should be way less flaky now.

smarterclayton commented:
/approve

Huang-Wei commented:
/lgtm
